Feat : JwtToken, EmailAuth 기능 추가

Dongjin113 opened the pull request:
### **_Feat : OW-46 CustomException 기능추가, ErrorCode 추가_**
- CustomException ErrorMessage String 문자열으로 생성자에 입력하면 자동으로 변경되게 변경
- EmailAuth 관련 ErrorCode 추가

### **_Feat : spring-boot-mail 의존성 추가_**
- javaMailSender 사용위한 spring-boot-mail 의존성 추가

### **_Feat OW-47 : UserEmailAuth Service 기능 구현_**

> ### TokenClaims 생성

- [x] JwtTokenConfig : 상속받은 Claims bean 으로 등록하여 사용할 설정 파일
- [x] 토큰관련 인증이 들어올 때 토큰관련 정보를 받아올 JwtTokenClaims 생성 한가지 추상 클래스로 만들어줄 Claims 에서 상속 받아 bean 등록한 후 한가지 타입으로 사용하기 위함 의존성 구분은 @qualifier 를 사용해 구분할 예정
- [x] JwtTokenClaims : 부모 추상클래스로 타입을 유연하게 사용하게위해 생성
- [x] JwtEmailAuthClaimsDto : 유저관련 토큰을 파싱하여 정보를 담을 Dto
- [x] EmailAuthClaimsDto : 이메일 인증관련 토큰을 파싱하여 얻은 정보를 담을 Dto

> ### Exception 생성
- [x] 유저가 존재하지 않을시 발생시킬 NotFoundUser Exception 생성자를 통해 메시지를 변경할 수 있도록 변경

> ### UserService
- [x] 유저의 회원가입 로직 생성 -> 추후 변경 예정

> ### Email 인증 관련 template 추가
- [x] email_template.html : 이메일 인증메일 template
- [x] email_auth_success.html : 이메일 인증 성공시 반환시켜줄 html template

> ### EmailAuthService
- [x] emailAuth : 이미 가입한 사용자인지 확인 후 인증관련 데이터(SSE를 위한 clientId, 인증하는 이메일) 저장 후 인증 메일 발송
- [x] 이메일 인증시 과거에 인증절차를 거친 사용자면 기존 데이터 삭제 후 재 등록
- [x] 이메일 인증시 이메일 토큰 생성 후 이메일 인증 메일에 url+token 값을 가진 인증버튼 전송
- [x] 이메일을 인증하거나 인증시간이 만료되면 SSE 연결을 끊고 요청이 들어온 클라이언트의 주소 삭제
- [x] 이메일 인증이 완료되면 DB에서 인증상태와 인증시간을 변경
- [x] 이메일 성공시 성공 메시지 클라이언트에 반환 후 Emitter 연결 종료


### **_Feat OW-48 : EmailAuth Security Config 구현_**
> ### EmailAuthenticationFilter
- [x] 요청이 들어오면 SecurityConfig에서 EmailAuthenticationFilter 실행이 된다
- [x] 이메일 인증 요청인지 확인한 다음
이메일 인증 요청이라면 처리하고 아니라면 인증 절차가 실행되지 않는다
- [x] 요청에서 넘어온 토큰이 유효 하다면 토큰을 파싱하여 얻은 이메일을 통해 존재하는 요청인지 확인 후 요청이 들어온 이메일에 대해 인증상태와 인증시간을 변경 후 클라이언트에 인증 완료 응답을 보낸다

> ### EmailAuthenticationToken
- [x] EmailAuthenticationFilter 로 들어온 요청의 토큰을 EmailAuthenticationToken 객체 로 생성한 후 AuthenticationManager 에 넘긴다

> ### AuthenticationManager
- [x] AuthenticationManager 에서 EmailAuthenticationToken 객체를 어떤 인증방식으로 처리할건지 정한 후 EmailAuthenticationProvider 토큰을 넘겨준다

> ### EmailAuthenticationProvider
- [x] 인증을 처리한후 데이터를 파싱하여 eamilAuthUserDetails 안에 값을 넣는다

> ### EmailAuthFailure
- [x] 이메일 인증이 실패한다면 실행될 Exception

> ### EmailAuthenticationToken
- [x] 이메일 인증을 통해 들어온 토큰을 인증할때 사용할 토큰 객체

> ### EmailAuthUserDetails
- [x] 이메일 인증을 통해 토큰에 대한 유저 정보가 담길 객체


### **_Feat OW-50 : Jwt Security Config 구현_**
> ### JwtAuthenticationFilter
사용자 로그인에 대한 필터이다
- [x] 요청이 들어오면 SecurityConfig에서 AuthenticationFilter 로 요청이 들어와 토큰 인증을 거쳐야하는지 확인한다
- [x] 토큰 인증을 거쳐야하는 요청이라면 Filter가 실행된다
- [x] Header에서 토큰값을 가져와서 JwtAuthenticationToken 객체에 토큰값을 담아서 AuthenticationManager에게 어떤 방식으로 인증을 진행할지 넘긴다
- [x] 모든 인증처리가 완료되면 SecurityContextHolder안에 사용자에 대한 정보가 담기고 @AuthenticationPrincipal
 Annotaion을 통해 호출해서 사용할 수 있다

> ### jwtAuthenticationToken
- [x] jwtAuthenticationToken 로 들어온 요청의 토큰을 jwtAuthenticationToken 객체 로 생성한 후 AuthenticationManager 에 넘긴다

> ### AuthenticationManager
- [x] AuthenticationManager 에서 JwtAuthenticationToken 객체를 어떤 인증방식으로 처리할건지 정한 후 JwtAuthenticationProvider 토큰을 넘겨준다

> ### JwtAuthenticationProvider
- 인증을 처리한후 데이터를 파싱하여 JwtUserDetails 안에 값을 넣는다

> ### JwtAuthFailure
- 사용자 인증이 실패한다면 실행될 Exception

> ### JwtAuthenticationToken
- 사용자 인증을 통해 들어온 토큰을 인증할때 사용할 토큰 객체

> ### JwtUserDetails
- 사용자 인증을 통해 토큰에 대한 유저 정보가 담길 객체

> ### JwtUtils
- 토큰에 대한 전반적인 내용을 담당한다
- [x] 토큰의 생성
- [x] 토큰의 인증
- [x] 토큰안에 들어갈 내용
- [x] 토큰의 생명 주기

> ### SecurityConfig
- [x] 각필터의 실행순서와 cors 권한 허용할 URL의 전반적인 처리를 맡는다
- [x] authenticationManager 가 각 토큰을 어떤 Provider를 통해 검증할 것인지 Bean 등록을 통해 설정해 준다
- [x] 각 필터에 필요한 객체를 Spring Container를 통해 주입될 수 있도록 Bean등록을 해준다

Dongjin113 committed:
Feat : OW-46 CustomException 기능추가, ErrorCode 추가

- CustomException ErrorMessage String 문자열으로 생성자에 입력하면 자동으로 변경되게 변경

- EmailAuth 관련 ErrorCode 추가

Dongjin113 committed:
Feat : spring-boot-mail 의존성 추가

- javaMailSender 사용위한 spring-boot-mail 의존성 추가

Dongjin113 committed:
Feat OW-47 : UserEmailAuth Service 기능 구현

### TokenClaims 생성
- [ ] JwtTokenConfig : 상속받은 Claims bean 으로 등록하여 사용할 설정 파일
- [ ] 토큰관련 인증이 들어올 때 토큰관련 정보를 받아올 JwtTokenClaims 생성 한가지 추상 클래스로 만들어줄 Claims 에서 상속 받아 bean 등록한 후 한가지 타입으로 사용하기 위함 의존성 구분은 @Qualifier 를 사용해 구분할 예정
- [ ] JwtTokenClaims : 부모 추상클래스로 타입을 유연하게 사용하게위해 생성
- [ ] JwtEmailAuthClaimsDto : 유저관련 토큰을 파싱하여 정보를 담을 Dto
- [ ] EmailAuthClaimsDto : 이메일 인증관련 토큰을 파싱하여 얻은 정보를 담을 Dto

### Exception 생성
- [ ] 유저가 존재하지 않을시 발생시킬 NotFoundUser Exception 생성자를 통해 메시지를 변경할 수 있도록 변경

### UserService
- [ ] 유저의 회원가입 로직 생성 -> 추후 변경 예정

### Email 인증 관련 template 추가
- [ ] email_template.html : 이메일 인증메일 template
- [ ] email_auth_success.html : 이메일 인증 성공시 반환시켜줄 html template

### EmailAuthService
- [ ] emailAuth : 이미 가입한 사용자인지 확인 후 인증관련 데이터(SSE를 위한 clientId, 인증하는 이메일) 저장 후 인증 메일 발송
- [ ] 이메일 인증시 과거에 인증절차를 거친 사용자면 기존 데이터 삭제 후 재 등록
- [ ] 이메일 인증시 이메일 토큰 생성 후 이메일 인증 메일에 url+token 값을 가진 인증버튼 전송
- [ ] 이메일을 인증하거나 인증시간이 만료되면 SSE 연결을 끊고 요청이 들어온 클라이언트의 주소 삭제
- [ ] 이메일 인증이 완료되면 DB에서 인증상태와 인증시간을 변경
- [ ] 이메일 성공시 성공 메시지 클라이언트에 반환 후 Emitter 연결 종료

Dongjin113 committed:
Feat OW-48 : EmailAuth Security Config 구현

### EmailAuthenticationFilter
- [ ] 요청이 들어오면 SecurityConfig에서 EmailAuthenticationFilter 실행이 된다
- [ ] 이메일 인증 요청인지 확인한 다음
이메일 인증 요청이라면 처리하고 아니라면 인증 절차가 실행되지 않는다
- [ ] 요청에서 넘어온 토큰이 유효 하다면 토큰을 파싱하여 얻은 이메일을 통해 존재하는 요청인지 확인 후 요청이 들어온 이메일에 대해 인증상태와 인증시간을 변경 후 클라이언트에 인증 완료 응답을 보낸다

### EmailAuthenticationToken
- [ ] EmailAuthenticationFilter 로 들어온 요청의 토큰을 EmailAuthenticationToken 객체 로 생성한 후 AuthenticationManager 에 넘긴다

### AuthenticationManager
- [ ] AuthenticationManager 에서 EmailAuthenticationToken 객체를 어떤 인증방식으로 처리할건지 정한 후 EmailAuthenticationProvider 토큰을 넘겨준다

### EmailAuthenticationProvider
- [ ] 인증을 처리한후 데이터를 파싱하여 eamilAuthUserDetails 안에 값을 넣는다

### EmailAuthFailure
- [ ] 이메일 인증이 실패한다면 실행될 Exception

### EmailAuthenticationToken
- [ ] 이메일 인증을 통해 들어온 토큰을 인증할때 사용할 토큰 객체

### EmailAuthUserDetails
- [ ] 이메일 인증을 통해 토큰에 대한 유저 정보가 담길 객체

Dongjin113 committed:
Feat OW-50 : Jwt Security Config 구현

### JwtAuthenticationFilter
사용자 로그인에 대한 필터이다
- [ ] 요청이 들어오면 SecurityConfig에서 AuthenticationFilter 로 요청이 들어와 토큰 인증을 거쳐야하는지 확인한다
- [ ] 토큰 인증을 거쳐야하는 요청이라면 Filter가 실행된다
- [ ] Header에서 토큰값을 가져와서 JwtAuthenticationToken 객체에 토큰값을 담아서 AuthenticationManager에게 어떤 방식으로 인증을 진행할지 넘긴다
- [ ] 모든 인증처리가 완료되면 SecurityContextHolder안에 사용자에 대한 정보가 담기고 @AuthenticationPrincipal
 Annotaion을 통해 호출해서 사용할 수 있다

### jwtAuthenticationToken
- [ ] jwtAuthenticationToken 로 들어온 요청의 토큰을 jwtAuthenticationToken 객체 로 생성한 후 AuthenticationManager 에 넘긴다

### AuthenticationManager
- [ ] AuthenticationManager 에서 JwtAuthenticationToken 객체를 어떤 인증방식으로 처리할건지 정한 후 JwtAuthenticationProvider 토큰을 넘겨준다

### JwtAuthenticationProvider
- 인증을 처리한후 데이터를 파싱하여 JwtUserDetails 안에 값을 넣는다

### JwtAuthFailure
- 사용자 인증이 실패한다면 실행될 Exception

### JwtAuthenticationToken
- 사용자 인증을 통해 들어온 토큰을 인증할때 사용할 토큰 객체

### JwtUserDetails
- 사용자 인증을 통해 토큰에 대한 유저 정보가 담길 객체

### JwtUtils
- 토큰에 대한 전반적인 내용을 담당한다
- [ ] 토큰의 생성
- [ ] 토큰의 인증
- [ ] 토큰안에 들어갈 내용
- [ ] 토큰의 생명 주기

### SecurityConfig
- [ ] 각필터의 실행순서와 cors 권한 허용할 URL의 전반적인 처리를 맡는다
- [ ] authenticationManager 가 각 토큰을 어떤 Provider를 통해 검증할 것인지 Bean 등록을 통해 설정해 준다
- [ ] 각 필터에 필요한 객체를 Spring Container를 통해 주입될 수 있도록 Bean등록을 해준다

dunowljj reviewed (approved):
작업량이 엄청나군요. 고생하셨습니다.

bstaran commented on back/src/main/java/com/ogjg/back/config/security/config/JwtTokenConfig.java:
```diff
@@ -0,0 +1,34 @@
+public class JwtTokenConfig {
+
+
+    @Bean
+    public JwtTokenClaims jwtUserClaimsDto() {
+        return new JwtUserClaimsDto();
+    }
+
+    @Bean
+    public JwtTokenClaims jwtEmailAuthClaims() {
+        return new JwtEmailAuthClaimsDto();
+    }
```
혹시 Dto를 빈으로 사용하시는 이유가 있나요?

Dongjin113 replied:
하나의 타입으로 좀 유연하게 쓸려고 시도해봤는데 
결국은 Qualifier로 무슨 빈을 사용하는지 명시해줘야해서 큰 의미는 없었던 것 같습니다
```

   /*
     * refreshToken 생성
     * */
    public String generateRefreshToken(
            @Qualifier("jwtUserClaimsDto") JwtTokenClaims userClaimsDto
    ) {
        return Jwts.builder()
                .setHeader(createHeader())
                .setClaims(createClaims(userClaimsDto))
                .setExpiration(new Date(System.currentTimeMillis() + REFRESH_TOKEN_VALID_TIME))
                .signWith(SIGNATURE_KEY)
                .compact();
    }

    /*
     * emailAuthToken 생성
     * */
    public String generateEmailAuthToken(
            @Qualifier("jwtEmailAuthClaims") JwtTokenClaims emailAuthClaimsDto
    ) {
        return Jwts.builder()
                .setHeader(createHeader())
                .setClaims(createClaims(emailAuthClaimsDto))
                .setExpiration(new Date(System.currentTimeMillis() + EMAIL_TOKEN_VALID_TIME))
                .signWith(SIGNATURE_KEY)
                .compact();
    }
```

bstaran replied:
이렇게 사용하면 Thread Safety하지 않지 않을까요?
제가 보기엔 동시성 문제가 발생할 가능성이 높아보입니다.ㅠㅠ

Dongjin113 replied:
수정했숨다

bstaran commented on back/src/main/java/com/ogjg/back/config/security/config/SecurityConfig.java:
```diff
@@ -0,0 +1,99 @@
+                .build();
+    }
+
+    @Bean
+    @Primary
+    public AuthenticationManager authenticationManager(
+            AuthenticationConfiguration authenticationConfiguration
+    ) throws Exception {
+        return authenticationConfiguration.getAuthenticationManager();
+    }
+
+    @Bean
+    public AuthenticationManager emailAuthenticationManager() throws Exception {
+        return new ProviderManager(Collections.singletonList(emailAuthenticationProvider()));
+    }
+
+    @Bean
+    public AuthenticationManager jwtAuthenticationManager() throws Exception {
+        return new ProviderManager(Collections.singletonList(jwtAuthenticationProvider()));
+    }
+
+    @Bean
+    public EmailAuthenticationFilter emailAuthenticationFilter() throws Exception {
+        return new EmailAuthenticationFilter(emailAuthenticationManager(), emailAuthService);
+    }
+
+    @Bean
+    public JwtAuthenticationFilter jwtAuthenticationFilter() throws Exception {
+        return new JwtAuthenticationFilter(jwtAuthenticationManager(), permitUrlList);
+    }
+
+    @Bean
+    public JwtAuthenticationProvider jwtAuthenticationProvider() {
+        return new JwtAuthenticationProvider(jwtUtils);
+    }
+
+    @Bean
+    public EmailAuthenticationProvider emailAuthenticationProvider() {
+        return new EmailAuthenticationProvider(jwtUtils);
+    }
```
각각의 빈으로 하나씩 등록한 이유가 있을까요?
각각의 필터 등록 메서드에서 설정을 하면 효율적이지 않을까요?

Dongjin113 replied:
오 좋은 방법 같습니다 수정해서 올리겠습니다

bstaran replied:
감사합니닷

bstaran commented on back/src/main/java/com/ogjg/back/config/security/emailauth/EmailAuthenticationFilter.java:
```diff
@@ -0,0 +1,72 @@
+        String uri = request.getRequestURI();
+
+        if (uri.startsWith("/favicon")) {
+//            todo favicon 어떻게 처리할지 고민해보기
```
👀

bstaran commented on back/src/main/java/com/ogjg/back/user/controller/UserController.java:
```diff
@@ -2,27 +2,53 @@
+    public ApiResponse<ImgUpdateResponse> updateImg(
+            @RequestParam("img") MultipartFile imgFile
+    ) {
         String loginEmail = "ogjg1234@naver.com";
```
이건 무슨 의미인가요?

dunowljj replied:
이건 제가 구현한 부분인데, 줄바꿈을 넣으신 듯 합니다.
- 이미지 데이터를 직접 받을때는 주로 `@RequestParam`에 `MultipartFiile`을 이용해서 받아야 한다고 합니다.
- 이 부분에 대해서는 아직 상세히 조사는 못했는데, 프론트에서 multipart/form-data 방식으로 보낸다해서 해당 형식과 맞추었습니다.

bstaran replied:
아 이렇게 표시가 됐는데, `String loginEmail = "ogjg1234@naver.com";`같은 코드들이 있는데 어떤 의미인가 궁금해서 여쭤봤습니다!
아직 구현이 안되어 있어서 테스트를 위해 사용하신건가요?

Dongjin113 committed:
Refactor OW-50 : Security Config 수정

- permitUrlList 요청 url user -> users로 변경
- AuthenticationManager Bean 등록 삭제 후 Filter 내부에서 명시 해줌

bstaran commented on back/src/main/java/com/ogjg/back/config/security/config/SecurityConfig.java:
```diff
@@ -0,0 +1,79 @@
+    private final List<String> permitUrlList = new ArrayList<>(
+            List.of("/api/users/email-auth/.*",
+                    "/signup",
+                    "health",
```
```suggestion
                    "/health",
```
경로 수정 부탁드립니다!

Dongjin113 replied:
수정했숨다

Dongjin113 committed:
Refactor OW-50 : Security Config 경로 수정, Clams 동시성 문제 개선

- JwtTokenConfig Spring Container에서 관리해주기 떄문에 Bean으로 등록하면 동시성 이슈가 발생할 가능성이 높기 때문에 들어오는 요청마다 새로은 Claims를 생성하게 해주기 위해 RequestScope annotation 사용

bstaran reviewed (approved):
고생하셨습니다!
